internal/proxy/accesslog/log_writer.go

```diff
@@ -0,0 +1,352 @@
+		return 0, fmt.Errorf(
+			"write length %d exceeds maximum file size %d", writeLen, l.max(),
+		)
+	}
+
+	if l.file == nil {
+		if err = l.openFileExistingOrNew(); err != nil {
+			return 0, err
+		}
+	}
+
+	if l.size+writeLen > l.max() {
+		if err := l.rotate(); err != nil {
+			return 0, err
+		}
+	}
+
+	n, err = l.file.Write(p)
+	l.size += int64(n)
+	return n, err
+}
+
+func (l *RotateLogger) Close() error {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	l.closeOnce.Do(func() {
+		close(l.closeCh)
+		if l.handler != nil {
+			l.handler.Close()
+		}
+
+		l.closeWg.Wait()
+	})
+
+	return l.closeFile()
+}
+
+func (l *RotateLogger) Rotate() error {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	return l.rotate()
+}
+
+func (l *RotateLogger) rotate() error {
+	if err := l.closeFile(); err != nil {
+		return err
+	}
+	if err := l.openNewFile(); err != nil {
+		return err
+	}
+	l.retention()
+	return nil
+}
+
+func (l *RotateLogger) openFileExistingOrNew() error {
+	l.retention()
+	filename := l.filename()
+	info, err := os.Stat(filename)
+	if os.IsNotExist(err) {
+		return l.openNewFile()
+	}
+	if err != nil {
+		return fmt.Errorf("file to get log file info: %s", err)
+	}
+
+	file, err := os.OpenFile(filename, os.O_APPEND|os.O_WRONLY, 0644)
+	if err != nil {
+		return l.openNewFile()
+	}
+
+	l.file = file
+	l.size = info.Size()
+	return nil
+}
+
+func (l *RotateLogger) openNewFile() error {
+	err := os.MkdirAll(l.dir(), 0744)
+	if err != nil {
+		return fmt.Errorf("make directories for new log file filed: %s", err)
+	}
+
+	name := l.filename()
+	mode := os.FileMode(0644)
+	info, err := os.Stat(name)
+	if err == nil {
+		mode = info.Mode()
+		newName := l.newBackupName()
+		if err := os.Rename(name, newName); err != nil {
+			return fmt.Errorf("can't rename log file: %s", err)
+		}
+		log.Info("seal old log to: " + newName)
+		if l.handler != nil {
+			l.handler.Update(newName, path.Base(newName))
+		}
+
+		// for linux
+		if err := chown(name, info); err != nil {
+			return err
+		}
+	}
+
+	f, err := os.OpenFile(name, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, mode)
+	if err != nil {
+		return fmt.Errorf("can't open new logfile: %s", err)
+	}
+	l.file = f
+	l.size = 0
+	return nil
+}
+
+func (l *RotateLogger) closeFile() error {
+	if l.file == nil {
+		return nil
+	}
+	err := l.file.Close()
+	l.file = nil
+	return err
+}
+
+func (l *RotateLogger) retentionRunOnce() error {
+	files, err := l.oldLogFiles()
+	if err != nil {
+		return err
+	}
+
+	if l.maxBackups > 0 && l.maxBackups < len(files) {
+		for _, f := range files[l.maxBackups:] {
+			errRemove := os.Remove(path.Join(l.dir(), f.fileName))
+			if err == nil && errRemove != nil {
+				err = errRemove
+			}
+		}
+	}
+
+	return err
+}
+
+func (l *RotateLogger) retentionRun() {
+	defer l.closeWg.Done()
+	for {
+		select {
+		case <-l.closeCh:
+			log.Warn("close Access log retention")
+			return
+		case <-l.retentionCh:
+			_ = l.retentionRunOnce()
+		}
+	}
+}
+
+func (l *RotateLogger) retention() {
+	select {
```

Comment:
format is wired

Reply:
> format is weird

like lumberjack, now I use mill instead.